yo1k/tic_tac_toe/test/core/test_game.py

```diff
@@ -0,0 +1,264 @@
+        self.assertEqual(game_state.board.cells, next_game_state_board)
+        self.assertEqual(game_state.step, 0)
+        self.assertEqual(game_state.round, 0)
+        self.assertEqual(game_state.phase, g.Phase.INROUND)
+        self.assertEqual(game_state.players[0].wins, 0)
+        self.assertEqual(game_state.players[1].wins, 0)
+
+
+class LogicQueueActionTest(unittest.TestCase):
+    def test_win_cur_round_0(self):
+        game_state = StateSetup._set_state(phase=g.Phase.ONSET)
+        act_queue_p1 = ListActionQueue([
+            g.Action.new_start(),
+            g.Action.new_occupy(g.Cell(1, 1)),
+            None,
+            g.Action.new_occupy(g.Cell(0, 0)),
+            g.Action.new_occupy(g.Cell(0, 2)),
+            g.Action.new_occupy(g.Cell(2, 0))])
+        act_queue_p2 = ListActionQueue([
+            g.Action.new_occupy(g.Cell(1, 2)),
+            g.Action.new_occupy(g.Cell(2, 2)),
+            None,
+            g.Action.new_occupy(g.Cell(0, 1))])
+        finish_game_state_board = [
+            [g.Mark.X, g.Mark.O, g.Mark.X],
+            [None, g.Mark.X, g.Mark.O],
+            [g.Mark.X, None, g.Mark.O]]
+        while game_state.phase is not g.Phase.OUTROUND:
+            g.Logic([act_queue_p1, act_queue_p2]).advance(game_state)
+        self.assertEqual(game_state.board.cells, finish_game_state_board)
+        self.assertEqual(game_state.step, 6)
+        self.assertEqual(game_state.round, 0)
+        self.assertEqual(game_state.phase, g.Phase.OUTROUND)
+        self.assertEqual(game_state.players[0].wins, 1)
+        self.assertEqual(game_state.players[1].wins, 0)
+
+    def test_draw_cur_round_1(self):
+        game_state = StateSetup._set_state(phase=g.Phase.INROUND, cur_round=1)
+        act_queue_p1 = ListActionQueue([
+            g.Action.new_occupy(g.Cell(0, 0)),
+            g.Action.new_occupy(g.Cell(1, 0)),
+            g.Action.new_occupy(g.Cell(0, 2)),
+            g.Action.new_occupy(g.Cell(2, 1))])
+        act_queue_p2 = ListActionQueue([
+            g.Action.new_occupy(g.Cell(1, 1)),
+            g.Action.new_occupy(g.Cell(1, 2)),
+            g.Action.new_occupy(g.Cell(2, 0)),
+            g.Action.new_occupy(g.Cell(0, 1)),
+            g.Action.new_occupy(g.Cell(2, 2))])
+        finish_game_state_board = [
+            [g.Mark.X, g.Mark.O, g.Mark.X],
+            [g.Mark.X, g.Mark.O, g.Mark.O],
+            [g.Mark.O, g.Mark.X, g.Mark.O]]
+        while game_state.phase is not g.Phase.OUTROUND:
+            g.Logic([act_queue_p1, act_queue_p2]).advance(game_state)
+        self.assertEqual(game_state.board.cells, finish_game_state_board)
+        self.assertEqual(game_state.step, 8)
+        self.assertEqual(game_state.round, 1)
+        self.assertEqual(game_state.phase, g.Phase.OUTROUND)
+        self.assertEqual(game_state.players[0].wins, 0)
+        self.assertEqual(game_state.players[1].wins, 0)
+
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
I remember you said that this is needed to run `test_game` as a module. But what if this code is removed, will you still be able to run the tests from `test_game` via unittest together with other tests? Will you be able to run only tests in `test_game` via unittest?

Reply:
> I remember you said that this is needed to run `test_game` as a module.

No, it is not. The purpose of that do not run the whole body of the module when you import it, but run all tests in the module when you run it separately.
If this code removed you still be able to run the tests from `test_game` via `unittest discover` together with other test. However, if you run `test_game` as python module tests will not run.

Reply:
Is it possible to specify some sort of a filter when running via unittest, so that only tests in `test_game` are run?

Reply:
Yes, it is possible if you use path to `test_game` only in `discover()`.
The other option is to use [`unittest.TestSuite()`](https://docs.python.org/3/library/unittest.html#:~:text=class%20unittest.TestSuite(tests%3D())%C2%B6).